fix: do not set max deadline for workspaces on template update

deansheather opened the pull request:
When templates are updated and schedule data is changed, we update all running workspaces to have up-to-date scheduling information that sticks to the new policy.

When updating the `max_deadline` for existing running workspaces, if the `max_deadline` was before `now()+2h` we would set the `max_deadline` to `now()+2h`.

Builds that don't/shouldn't have a `max_deadline` have it set to `0`, which is always before `now()+2h`, and thus would always have the `max_deadline` updated.

Closes https://github.com/coder/customers/issues/530

deansheather committed:
fix: do not set max deadline for workspaces on template update

When templates are updated and schedule data is changed, we update all
running workspaces to have up-to-date scheduling information that sticks
to the new policy.

When updating the max_deadline for existing running workspaces, if the
max_deadline was before now()+2h we would set the max_deadline to
now()+2h.

Builds that don't/shouldn't have a max_deadline have it set to 0, which
is always before now()+2h, and thus would always have the max_deadline
updated.

deansheather committed:
fixup! fix: do not set max deadline for workspaces on template update

sreya reviewed (commented):
Looks good but I'm going to wait for tests from @Emyrk before approving 👍

deansheather committed:
fixup! fix: do not set max deadline for workspaces on template update

Emyrk committed:
test: add unit test to excercise template schedule bug

added a user without quiet hours and a template without a autostop
requirment to trigger the bug.

Emyrk committed:
add test comment

Emyrk committed:
add unit test to ensure deadline is never nil if maxdeadline is set

Emyrk committed:
add comment to explain grace period

Emyrk reviewed (approved):
I've added unit tests that fail on `main` reproducing the bug. 

In practice (I have not manually confirmed), I think these are the steps to reproduce.

1. Template autostart requirements must be disabled. [This if condition needs to be falsy](https://github.com/coder/coder/blob/295b7557260b3a7942c75c241212ddb484599b95/coderd/schedule/autostop.go#L120-L120)
2. You need to have a workspace that has a deadline beyond 2hrs in the future using said template.
3. Update the template schedule (use max ttl should be off)
4. Workspace deadline should now be 2hrs after the template `updated_at`

Test: https://github.com/coder/coder/blob/295b7557260b3a7942c75c241212ddb484599b95/enterprise/coderd/schedule/template_test.go#L135-L154
